src/Network/ABCI/Types/App.hs

```diff
@@ -1,21 +1,89 @@
-module Network.ABCI.Types.App
-  ( App(..)
-  , transformApp
-  ) where
 
+module Network.ABCI.Types.App where
+
+import           Control.Lens                         ((^.))
+import           Control.Monad                        ((>=>))
+import           Data.Bifunctor                       (first)
+import qualified Data.ByteString                      as BS
+import           Data.Function                        ((&))
+import qualified Data.ProtoLens                       as PL
+import           Data.ProtoLens.Encoding.Bytes        (getVarInt, putVarInt,
+                                                       runBuilder, runParser,
+                                                       signedInt64ToWord,
+                                                       wordToSignedInt64)
+import           Data.String.Conversions              (cs)
+import           Data.Text                            ()
+import           Data.Traversable                     (traverse)
+import           Network.ABCI.Types.DecodeError             (DecodeError)
+import qualified Network.ABCI.Types.DecodeError             as DecodeError
 import           Network.ABCI.Types.Messages.Request  (Request)
+import qualified Network.ABCI.Types.Messages.Request  as Request
 import           Network.ABCI.Types.Messages.Response (Response)
+import qualified Network.ABCI.Types.Messages.Response as Response
 import           Network.ABCI.Types.Messages.Types    (MessageType)
+import qualified Proto.Types                          as PT
+import qualified Proto.Types_Fields                   as PT
 
 -- | Application type that represents a well typed application, i.e. a
 -- function from a typed `Request` to a typed `Response`.
 newtype App m = App
-  { runApp :: forall (t :: MessageType). Request t -> m (Response t) }
+  { unApp :: forall (t :: MessageType). Request t -> m (Response t) }
 
 -- | Transform an application from running in a custom monad to running in `IO`.
-transformApp
-  :: (forall a. m a -> n a)
-  -> App m
-  -> App n
-transformApp nat (App f) = App $ \req ->
-  nat $ f req
+transformApp :: (forall a. m a -> g a) -> App m -> App g
+transformApp nat (App f) = App $ nat . f
+
+-- | compiles `App` down to `AppBS`
+runApp :: Applicative m => App m -> AppBS m
+runApp (App app) bs =
+  bs
+    & (decodeLengthPrefix >=> decodeRequests)
+    & either (pure . onError) (traverse onResponce)
+    & fmap (encodeLengthPrefix . encodeResponces)
+  where
+    onError :: DecodeError -> [PT.Response]
+    onError err = [Response.toProto $ Response.ResponseException $ Response.Exception $ cs $ DecodeError.print err]
+    -- for some reason with this type we get • Couldn't match type ‘m1’ with ‘m’
+    -- onResponce :: PT.Request'Value -> m PT.Response
```

Comment:
Any ideas on this?

Reply:
You could try passing `App m` in an argument to `onResponse` in the where clause and pattern match there. 

```haskell

onResponce :: App m -> PT.Request'Value -> m PT.Response


```

you might rename it to `buildResponse` or something, idk

Reply:
Adding `forall m.` to the definition of `runApp` fixed the issue